torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -596,9 +580,90 @@ def _update_trainable(self) -> None:
         # Create the optim which will work on the param shard
         if not hasattr(self, "optim"):
             self._clear_cache()
-            self._default_device = list(self._per_device_params.keys())[0]
             self.optim = self._optim_constructor(self.partition_parameters()[self.rank], **self._optim_defaults)
             self._sync_param_groups(self.optim.param_groups, self.param_groups)
 
         if self.parameters_as_bucket_view:
             self._setup_flat_buffers()
+
+    def _verify_and_init_params(self, params: Any) -> None:
+        r"""
+        Verifies the type of ``params`` and initializes ``self._all_params``
+        if ``params`` is valid.
+
+        While :class:`optim.Optimizer <torch.optim.Optimizer>` allows
+        ``params`` to be an iterable of :class:`dict` s, currently
+        ``ZeroRedundancyOptimizer`` strictly requires ``params`` to be an
+        iterable of :class:`torch.Tensor` s.
+
+        Raises:
+            TypeError: ``params`` has an invalid type.
+            ValueError: ``params`` is empty.
+        """
+        if (isinstance(params, torch.Tensor)):
+            raise TypeError("params argument should be an iterable of "
+                            f"Tensors, but got {torch.typename(params)}")
+        try:
+            self._all_params = list(params)
+        except TypeError:
+            raise TypeError("params argument should be an iterable of "
+                            f"Tensors, but got {torch.typename(params)}")
+        if (len(self._all_params) == 0):
+            raise ValueError("ZeroRedundancyOptimizer got an empty parameter "
+                             "list")
+        for param in self._all_params:
+            if (not isinstance(param, torch.Tensor)):
+                raise TypeError("params argument should be an iterable of"
+                                "Tensors, but got an iterable containing "
+                                f"{torch.typename(param)}")
+
+    def _verify_same_param_device(self) -> None:
+        r"""
+        Verifies that ZeRO is being used under the single-process single-
+        device regime where a process operates exclusively on a full model
+        replica on a single device.
+
+        The function assumes that ``self._all_params`` has been initialized
+        and is non-empty.
+
+        Raises:
+            ValueError: ``params`` contains parameters across multiple
+                devices.
+
+        NOTE: This function can be removed once support for sharding a rank's
+        model parameters across multiple devices is added.
+        """
+        device = self._all_params[0].device
+        for param in self._all_params[1:]:
+            if (param.device != device):
+                raise ValueError("ZeroRedundancyOptimizer assumes that each "
+                                 "rank's model parameters are on the same "
+                                 f"device but got both {str(device)} and "
```

Comment:
curious, is `str` necessary? I had an impression that `device` does have `__str__` or `__repr__` methods implemented.

Reply:
Good catch. It was a remnant from switching to using f-strings. Fixed.